### ISSUE TITLE
Sweeps: Conditional parameters

### ISSUE THREAD

**Issue by EliaCereda** [wandb/wandb#1487]
It seems to me that there is currently no way to define a sweep where some parameters depends on another one, is it correct?

What I have in mind is something like this example from `hyperopt` ([source](https://github.com/hyperopt/hyperopt/wiki/FMin#22-a-search-space-example-scikit-learn)):

```python
from hyperopt import hp
space = hp.choice('classifier_type', [
    {
        'type': 'naive_bayes',
    },
    {
        'type': 'svm',
        'C': hp.lognormal('svm_C', 0, 1),
        'kernel': hp.choice('svm_kernel', [
            {'ktype': 'linear'},
            {'ktype': 'RBF', 'width': hp.lognormal('svm_rbf_width', 0, 1)},
        ]),
    },
    {
        'type': 'dtree',
        'criterion': hp.choice('dtree_criterion', ['gini', 'entropy']),
        'max_depth': hp.choice('dtree_max_depth',
            [None, hp.qlognormal('dtree_max_depth_int', 3, 1, 1)]),
        'min_samples_split': hp.qlognormal('dtree_min_samples_split', 2, 1, 1),
    },
])
```

At the moment this is not support by neither the "standard" sweeps nor by the beta Ray Tune sweeps, right?

**Comment by cvphelps** [comment]
Hi Elia, right now this is not well supported (have you tried the ray tune beta with this? do you get an error?) It's a good idea and we'll add to the backlog of tickets.

**Comment by EliaCereda** [comment]
With the Ray Tune beta, if you create a sweep with a configuration such as the above, everything is set up correctly. When you start the agent, though

**Comment by EliaCereda** [comment]
This gist produces a sweep configuration that demonstrates the issue: https://gist.github.com/EliaCereda/149f125232e6d53e9eca83bc91608047

Steps to reproduce:
1. Create a sweep with `python raytune-nested-test.py`
2. Start the agent using the ID returned by the script `wandb agent #####`
3. See the command line arguments supplied to the script

```
/usr/bin/env python example.py "--hyperopt.hp.choice=['classifier_type', [{'type': 'naive_bayes'}, {'C': {'hyperopt.hp.uniform': ['svm_C', 0, 1]}, 'kernel': {'hyperopt.hp.choice': ['svm_kernel', [{'ktype': 'linear'}, {'ktype': 'RBF', 'width': {'hyperopt.hp.uniform': ['svm_rbf_width', 0, 1]}}]]}, 'type': 'svm'}, {'criterion': {'hyperopt.hp.choice': ['dtree_criterion', ['gini', 'entropy']]}, 'max_depth': {'hyperopt.hp.choice': ['dtree_max_depth', [None, {'hyperopt.hp.uniform': ['dtree_max_depth_int', 3, 1, 1]}]]}, 'min_samples_split': {'hyperopt.hp.uniform': ['dtree_min_samples_split', 2, 1, 1]}, 'type': 'dtree'}]]"
```

**Comment by ariG23498** [comment]
Hey @EliaCereda 
Thanks for the gist
That will help in tracking the issues mentioned by you. :smile:

**Comment by emanuelevivoli** [comment]
hi, any news?

**Comment by itchingpixels** [comment]
hi, we'd love to see this as well!

**Comment by jhuebotter** [comment]
Would love to see this as well :)

**Comment by nikikilbertus** [comment]
Would love to see this as well!

**Comment by aapowadi** [comment]
I would love to see this feature as well.

**Comment by pdeubel** [comment]
Any updates on this? I would love to see this as well!

**Comment by emilysilcock** [comment]
Would also love to see this!

**Comment by lesliewandb** [comment]
I have bumped it up to our team again

**Comment by Matthijs-Kemp** [comment]
Same here, this feature would be really useful to our team.

**Comment by lesliewandb** [comment]
Thank you for letting us know about how much this feature is desired. I've been letting our team know about this and will let you guys know when we have enough resources to implement this

**Comment by Pragalbhv** [comment]
This feature would make a lot of our work easier! Looking forward to this!

**Comment by amariucaitheodor** [comment]
Also looking forward to this!

**Comment by zimmer-m** [comment]
Is there any chance this will be implemented anytime soon? the absence of this feature makes my workflow much slower

**Comment by EchteRobert** [comment]
Would love to see this implemented!

**Comment by Hauserrodr** [comment]
I would love this feature, there's a lot of cases where you need to have conditions on parameters. One example is to choose a patch size that the image size is divisible by, in image transformers networks. There's lot of examples with feature extraction using spectrograms for audio too.

**Comment by ptempelman** [comment]
Would also appreciate this!

**Comment by mmcdermott** [comment]
This would be extremely useful!

**Comment by mmcdermott** [comment]
@cvphelps Any update on this feature? It would be extremely useful, and a timetable would be greatly appreciated.

**Comment by Franz28** [comment]
Hi, I'm here to show support for this cause!
Conditional parameters also make sense when architectures are part of the hyperparameters to be sweeped (eg. comparing different approaches inside the same sweep).

**Comment by mmcdermott** [comment]
Does anyone know if contributions are welcome on this issue? I think the necessary code changes would mostly live in this repo: https://github.com/wandb/sweeps 
and it may not bee too involved to implement this functionality...

**Comment by rsanandres-wandb** [comment]
As of right now and after consideration, the team has put the feature on hold and there is no current plans for implementation for now. However, with a beta Launch feature you are able to use Optuna as the optimizer which does support conditional configs. ([Guide in the Docs](https://docs.wandb.ai/guides/launch/sweeps-on-launch#create-a-custom-sweep-scheduler)).

**Comment by mmcdermott** [comment]
@rsanandres-wandb , are you open to contributions in the https://github.com/wandb/sweeps by the community on this issue?

**Comment by sydholl** [comment]
WandB Internal User commented: 
Hauserrodr commented: 
I would love this feature, there's a lot of cases where you need to have conditions on parameters. One example is to choose a patch size that the image size is divisible by, in image transformers networks. There's lot of examples with feature extraction using spectrograms for audio too.

**Comment by sydholl** [comment]
WandB Internal User commented: 
Franz28 commented: 
Hi, I'm here to show support for this cause!
Conditional parameters also make sense when architectures are part of the hyperparameters to be sweeped (eg. comparing different approaches inside the same sweep).

**Comment by sydholl** [comment]
WandB Internal User commented: 
rsanandres-wandb commented: 
As of right now and after consideration, the team has put the feature on hold and there is no current plans for implementation for now. However, with a beta Launch feature you are able to use Optuna as the optimizer which does support conditional configs. ([Guide in the Docs](https://docs.wandb.ai/guides/launch/sweeps-on-launch#create-a-custom-sweep-scheduler)).

**Comment by sydholl** [comment]
WandB Internal User commented: 
mmcdermott commented: 
Does anyone know if contributions are welcome on this issue? I think the necessary code changes would mostly live in this repo: https://github.com/wandb/sweeps 
and it may not bee too involved to implement this functionality...

**Comment by tristan-spaulding2** [comment]
Hi--as mentioned, this is now possible to do with W&B's Sweeps on Launch feature.  Here are two notebooks getting it up and running:  

- [Setting up Launch](https://colab.research.google.com/drive/1wX0OSVxZJDHRsZaOaOEDx-lLUrO1hHgP#scrollTo=jhm3qUUxk69o)
- [Conditional configs and multi-objective optimization](https://colab.research.google.com/drive/1WxLKaJlltThgZyhc7dcZhDQ6cjVQDfil#scrollTo=sHcr30CKybN7)

Beyond these features, Launch also offers some better scalability and parallelization patterns for Sweeps (and normal training jobs).

**Comment by jtwild** [comment]
I would also be interested in having this feature in the standard wandb sweeps!

**Comment by sydholl** [comment]
This feature was implemented as of April 2024

**Comment by Guill-Hub** [comment]
Hi, I couldn't find where to search in the documentation to use this. Does anyone know?

**Comment by jhuebotter** [comment]
I am also at a loss here. Some more documentation / tutorial for this feature (if it does exist?) would be much appreciated.

**Comment by n00b001** [comment]
@sydholl this is marked as closed / completed. How do we use this feature?
